### ISSUE TITLE
Same script, multiple concurrent instances, on Linux (again)

### ISSUE THREAD

**Issue by rconde01** [oleg-shilo/cs-script#67]
Hey Oleg,
  I spent some time finally to dig into this issue and experiment a bit. Here is my test:

cs_script_test.cs
```
using System;

class A
{
   static void Main(string [] args)
   {
      Console.WriteLine("boo!!!");
   }
}
```

test1.sh
```
#!/bin/sh

cscs.exe -cache:clear

cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
cscs.exe -nl cs_script_test.cs &
wait 
```

test2.sh
```
#!/bin/sh

cscs.exe -cache:clear

cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
cscs.exe -nl -inmem:0 cs_script_test.cs &
wait 
```

On redhat 7 with mono 5 test1 and test2 randomly crash. 

I was able to fix it by changing csscript.cs from:

`using(SystemWideLock compilingFileLock = new SystemWideLock(options.scrptFileName,null))`

to

`using(SystemWideLock compilingFileLock = new SystemWideLock(options.scrptFileName,"c"))`

I think the underlying issue is described here:
http://man7.org/linux/man-pages/man2/fcntl.2.html

> If a process closes any file descriptor referring to a file, then all of the process's locks on that file are released, regardless of the file descriptor(s) on which the locks were obtained.  This is bad: it means that a process can lose its locks on a file such as /etc/passwd or /etc/mtab when for some reason a library function decides to open, read, and close the same file.

Thus, perhaps, by using a unique file name which (shouldn't) be accessed elsewhere in the application, you avoid inadvertently releasing the lock.

On a related topic, in that section of the code you have:

`lockedByCompiler = !compilingFileLock.Wait(3000); `

I don't think you want the !...although the variable is never used so the function is the same.

Also, I'm not sure about the 3000 ms timeout. Compilation would hopefully be less than 3 seconds, but if you have a ton of processes waiting on the lock maybe it's not unrealistic to wait that long. I'm also not sure if there's any guarantee regarding the order in which processes get the lock. So for all I know the first waiter could end up running last. This is not a scenario I've run into but I figured it was worth bringing up.

**Comment by oleg-shilo** [comment]
Great. Thank you I will have a look at it. Unfortunately file lock on Linux is not that straight forward. And absence of a system wide Mutex on Mono/Linux is not helping at all.

As for the `lockedByCompiler = !compilingFileLock.Wait(3000);` the lockedByCompiler is a local unused var just for debugging/troubleshooting.  And the variable it means locked by some **other** compiler call (wait returned false) so `!` is kind of intended.

3000 is truly arbitrary. And in case of lock it is how much I want to wait before giving up and throwing the exception. And if the resource becomes available before 3000 expires then the execution will resume. 3000 is also kind of comparable with some heavy compilers first loading time. Like Roslyn ~3 sec on Win and ~1.4 on Linux.    

And I will investigate the `null`->`"c"` effect. Txs.

**Comment by oleg-shilo** [comment]
Your proposal has been incorporated with 5cd04a4.
Thank you.

**Comment by rconde01** [comment]
Just to clarify...the "unintended release" idea was just a guess...I didn't confirm it...but in any case the "c" fix DID fix the problem...

**Comment by oleg-shilo** [comment]
Not a problem. I am just lost when I try to answer how adding an extra constant suffix for the file lock name changes the things but... I am happy that it does.... :)
...
I just had a small revelation moment. It actually see now one small possibility for that suffix to do the trick....

Without the suffix the lock name is the name of the file that **does** exist on the system but with the suffix it is just a "virtual file". Meaning that the system may handle these lock differently. Even if we don't understand how.

Before this very moment I was convinced that the names of the locks are all different but all three share the same nature. And now I realize that I was wrong. One name had a special meaning - it had an artifact (a file) associated with it, while other two were pure "virtual".

The change you proposed is 100% safe as it is what Win Mutex does. It uses arbitrary names for kernel synchronization objects. For the Linux equivalent I decided to recycle one name (script file name). It was no real reason to do that and it was no compelling reason not to do that. So I did. But now I have a very good reason to stop using raw file name and start useing the suffix -  yours _"in any case the "c" fix DID fix the problem..."_ .  So I am happily going with your fix.
